### PR TITLE
open reader only once with correct schema

### PR DESCRIPTION
**Pull request by robert3005**
This should have been like that from the start when we bumped parquet.

### CODE REVIEW

**Comment by robert3005** [comment]
@mccheah @ash211 @pwoody @sjrand might have found something. Also removing duplicate opens which are bad on s3

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/SpecificParquetRecordReaderBase.java [hunk 1]
```diff
@@ -106,13 +106,12 @@ public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptCont
       // then we need to apply the predicate push down filter
       footer = readFooter(configuration, file, range(split.getStart(), split.getEnd()));
       FilterCompat.Filter filter = getFilter(configuration);
-      ParquetFileReader reader = ParquetFileReader.open(taskAttemptContext.getConfiguration(), file, footer);
+      reader = ParquetFileReader.open(taskAttemptContext.getConfiguration(), file, footer);
```
Comment (schlosna) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/SpecificParquetRecordReaderBase.java]: nit: `this.reader`

Comment (ash211) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/SpecificParquetRecordReaderBase.java]: just use `configuration` ?  This is a different technique for creating the ParquetFileReader than below -- why is that?

**Comment by robert3005** [comment]
We don't since apache parquet still hasn't published a release. Upstream is still on parquet 1.7 and this is 1.9 change.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/SpecificParquetRecordReaderBase.java [hunk 2]
```diff
@@ -141,17 +140,18 @@ public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptCont
...
       }
+      this.reader = new ParquetFileReader(configuration, file, footer);
     }
     this.fileSchema = footer.getFileMetaData().getSchema();
     Map<String, String> fileMetadata = footer.getFileMetaData().getKeyValueMetaData();
     ReadSupport<T> readSupport = getReadSupportInstance(getReadSupportClass(configuration));
     ReadSupport.ReadContext readContext = readSupport.init(new InitContext(
         taskAttemptContext.getConfiguration(), toSetMultiMap(fileMetadata), fileSchema));
     this.requestedSchema = readContext.getRequestedSchema();
+    reader.setRequestedSchema(requestedSchema);
```
Comment (ash211) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/SpecificParquetRecordReaderBase.java]: This is part of using the new Parquet version?  What does it do?

**Comment by schlosna** [comment]
:+1: LGTM

@robert3005 gotcha, should we create an issue here to track all the relevant changes for Parquet bump?

**Comment by robert3005** [comment]
@ash211 turns out there's one https://issues.apache.org/jira/browse/SPARK-13127

**Comment by ash211** [comment]
Filed pushing the upgrade upstream as https://github.com/palantir/spark/issues/60 so we don't lose track of it.  Any other future Parquet things we change should be added to that ticket so we can eventually get it into upstream Spark

**Comment by robert3005** [comment]
Cool, thanks for the ticket.

**Comment by pwoody** [comment]
👍

**Pull request merged**